doc/proposals/keynames.md

```diff
@@ -0,0 +1,226 @@
+
+### Storage Plugins
+
+Storage plugins for flat formats (without a native nesting hierarchy) should just use the `keyName` to store values (and metadata, if possible).
+
+Storage plugins with support for a native hierarchy should use the operations `store` and `load` to find the names to use in that hierarchy (e.g. field names in JSON).
+
+`S` is a subset of `R`.
+It is the set of reserved characters, whose meaning has a native representation in the plugins storage format.
+e.g. `#` is in `S`, if the storage format has native arrays.
+`\` must not be in `S`.
+
+The following operations would be provided in a separate library for storage plugins:
+
+- `store(S, u)`: turns an unescaped part `u` into a name the storage plugin should use.
+  Defined as:
+  - If `u[0] = \` and `u[1]` is in `S` then `store(S, u) := u[1:]`
+  - If `u[0] = \` and `u[1]` is not in `S` then `store(S, u) := u`.
+    (NB:`u[1]` must be in `R` because `u` is valid).
+  - If `u[0]` is in `S` then `store (S, u) := NULL`.
+    Indicates to the plugin that it should do its native handling.
+  - Otherwise `store(S, u) := u`.
+- `load(k, S, n)`: adds an unescaped part corresponding to the name `n` (produced by `store`) to the key `k`.
+  Defined as:
+  - If `n[0]` is in `S` then `load(k, S, n) := pushL(k, n)`.
+  - Otherwise `load(k, S, n) := pushU(k, n)`.
+
+#### Directory values
+
+Directory keys with values can be handled much more cleanly.
+
+I will use `\value` in the example, but any other invalid unescaped key part whose first character is not in `S` can also be used.
+
+I will use JSON as an example, it requires that a `\` is escaped as `\\`.
+
+##### Store procedure
+
+If there are keys below the key `k`, store the value of `k` in the nested field `\value`:
+
+```json
+"k": {
+    "\\value": "value of k",
+    ...
+}
+```
+
+Otherwise, store the value directly.
+
+```json
+"k": "value of k"
+```
+
+If `k` is an array parent, I recommend introducing a pseudo-level in the hierarchy (other ways might work to).
+In JSON for example:
+
+```json
+"k": {
+    "\\value": "value of k",
+    "\\array": [ ... ]
+}
+```
+
+The benefit of this solution is, it also works for mixed array and non-array parents.
+However, I wouldn't generally recommend using such structures.
+
+##### Load procedure
+
+If we encounter a field named `\value`, we store its value not in `p/\value` (which is invalid anyway), but instead in the parent key `p`.
+
+If we encounter the `\array` field from above, we ignore it when constructing names for the elements of the array.
+We don't construct `p/\array/#1` (again invalid), but `p/#1`.
+
+#### Metadata
+
+A similar construction to the `\array` handling above, but instead with a nested `\meta` object can be used for metadata as well.
+
+For example a directory key with value and metadata in JSON:
+
+```json
+{
+"key": {
+    "\\value": "value of key",
+    "\\meta": {
+        "type": "string"
+    },
+    "subkey": 1
```

Comment:
> This would implicitly create metadata `type=long`.

Please don't assume anything based on knowledge of existing plugins here. In an actual JSON plugin this should create `type=long` metadata yes. The plugin also shouldn't store the `type` metadata of strings, since it is redundant. But again, I wasn't trying to show, how a JSON plugin should work. I only explained how directory keys, metadata and values can be stored together in a format without native support for that. JSON is used, only because it is easy to read and well known.